### PR TITLE
Refactor CLIs for merging sharded files

### LINKED ISSUE #1160: Users should be able to set size of buffer used for merging files
We expose this as a parameter on [`FileMerger`](https://github.com/bigdatagenomics/adam/blob/master/adam-core/src/main/scala/org/bdgenomics/adam/rdd/FileMerger.scala#L48) but don't actually give anyone the hooks to change it anywhere.

### PR DESCRIPTION
**Pull request by fnothaft**
Resolves #1160, #1161, #1165.

### CODE REVIEW

File: adam-cli/src/main/scala/org/bdgenomics/adam/cli/Fragments2Reads.scala [hunk 1]
```diff
@@ -43,6 +43,8 @@ class Fragments2ReadsArgs extends Args4jBase with ADAMSaveAnyArgs with ParquetAr
   var asSingleFile: Boolean = false
   @Args4jOption(required = false, name = "-sort_reads", usage = "Sort the reads by referenceId and read position")
   var sortReads: Boolean = false
+  @Args4jOption(required = false, name = "-deferMerging", usage = "Defers merging single file output")
```
Comment (heuermh) [inline adam-cli/src/main/scala/org/bdgenomics/adam/cli/Fragments2Reads.scala]: `-camelCase` or `-snake_case`?

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/AlignmentRecordRDD.scala [hunk 2]
```diff
@@ -405,11 +410,14 @@ sealed trait AlignmentRecordRDD extends AvroReadGroupGenomicRDD[AlignmentRecord,
...
-        tailPath,
-        optHeaderPath = Some(headPath),
-        writeEmptyGzipBlock = !asSam)
+      if (!deferMerging) {
+        FileMerger.mergeFiles(rdd.context.hadoopConfiguration,
+          fs,
+          outputPath,
+          tailPath,
+          optHeaderPath = Some(headPath),
+          writeEmptyGzipBlock = !asSam)
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/AlignmentRecordRDD.scala]: will this boolean `writeEmptyGzipBlock` flag still work ok when CRAM is added?

**Comment by fnothaft** [comment]
Just added some tests. This is good to go from my side. Can I get a review/merge? Also, let me know if people would like ba2c484 squashed down.

File: adam-cli/src/main/scala/org/bdgenomics/adam/cli/MergeShards.scala [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+class MergeShardsArgs extends Args4jBase {
+  @Argument(required = true, metaVar = "INPUT", usage = "The shard directory to merge", index = 0)
+  var inputPath: String = null
+  @Argument(required = true, metaVar = "OUTPUT", usage = "The location to write the merged file", index = 1)
+  var outputPath: String = null
+  @Args4jOption(required = false, name = "-header_path", usage = "Optional path to a header")
+  var headerPath: String = null
+  @Args4jOption(required = false,
+    name = "-bufferSize",
```
Comment (heuermh) [inline adam-cli/src/main/scala/org/bdgenomics/adam/cli/MergeShards.scala]: sorry, another `-camelCase` to `-snake_case`

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/FileMerger.scala [hunk 4]
```diff
@@ -41,23 +47,68 @@ private[rdd] object FileMerger extends Logging {
...
+                                                tailPath: Path,
+                                                optHeaderPath: Option[Path] = None,
+                                                writeEmptyGzipBlock: Boolean = false,
+                                                writeCramEOF: Boolean = false,
+                                                optBufferSize: Option[Int] = None) {
+
+    // check for buffer size in option, if not in option, check hadoop conf,
+    // if not in hadoop conf, fall back on 4MB
+    val bufferSize = optBufferSize.getOrElse(conf.getInt(BUFFER_SIZE_CONF,
+      4 * 1024 * 1024))
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/FileMerger.scala]: can this default be mentioned at the command line above, say
```
usage = "Buffer size for merging single file output, default 4MB"
```

Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/FileMerger.scala]: SGTM! I've addressed this in `MergeShardsArgs`.

**Comment by fnothaft** [comment]
Rebased and addressed review comments.

**Comment by heuermh** [comment]
Nice new Github review feature, don't have to leave LGTM comments any more.  Well, except there's this one.

**Comment by heuermh** [comment]
Merged as multiple commits.

Thank you, @fnothaft!

**Pull request merged**